src/main/java/org/craftercms/engine/service/context/SiteContextManager.java

```diff
@@ -303,6 +340,48 @@ public void startContextRebuild(String siteName, boolean fallback, Consumer<Site
         });
     }
 
+    /**
+     * Get a site context and initializing if it does not exist.
+     * If in preview mode, do not retry on failure.
+     * If in live mode, retry on failure. After the max retries count, throw an exception.
+     * @param siteName the site name of the context
+     * @return the site context
+     * @throws InterruptedException if the current thread is interrupted while waiting
+     */
+    protected SiteContext getContextWithInitializing(String siteName) throws InterruptedException {
+        if (modePreview) {
+            try {
+                // If the site context doesn't exist (it's new), it will be created
+                return getContext(siteName, false);
+            } catch (Exception e) {
+                logger.error(format("Error creating site context for site '%s'", siteName), e);
+            }
+
+            return null;
+        }
+
+        int attempt = 0;
+        long waitTime = contextBuildRetryWaitTimeBase;
+        while (attempt < contextBuildRetryMaxCount) {
+            try {
+                // If the site context doesn't exist (it's new), it will be created
+                return getContext(siteName, false);
+            } catch (Exception e) {
+                attempt++;
+                if (attempt == contextBuildRetryMaxCount) {
+                    logger.error(format("Error creating site context for site '%s'", siteName), e);
+                    throw e;
+                }
+
+                logger.warn(format("Error creating site context for site '%s'. Retrying in '%d' seconds", siteName, waitTime / 1000), e);
+                contextBuildLatch.await(waitTime, TimeUnit.MILLISECONDS);
```

Comment:
We don't need to use a countdown latch here, since this method is only called on startup by a single thread

Reply:
Fixed to just use `Thread.sleep`. Thank you!